### PR TITLE
Fixed 31235 -- Updated assertQuerysetEqual to compare querysets directly

### PR DESCRIPTION
**Pull request by inglesp**

### CODE REVIEW

File: django/test/testcases.py [hunk 1]
```diff
@@ -1030,8 +1030,18 @@ def _fixture_teardown(self):
...
+                if isinstance(values[0], str):
+                    transform = repr
+                else:
+                    transform = lambda o: o  # NOQA
+
         items = map(transform, qs)
+
         if not ordered:
             return self.assertEqual(Counter(items), Counter(values), msg=msg)
         values = list(values)
```
Comment (charettes) [inline django/test/testcases.py]: This can be removed now that `values = list(values)` is done above.

File: django/test/testcases.py [hunk 2]
```diff
@@ -1030,8 +1030,18 @@ def _fixture_teardown(self):
...
+
+        if values:
+            if transform is None:
+                if isinstance(values[0], str):
+                    transform = repr
+                else:
+                    transform = lambda o: o  # NOQA
+
         items = map(transform, qs)
+
```
Comment (charettes) [inline django/test/testcases.py]: Unnecessary whitespace change.

File: django/test/testcases.py [hunk 3]
```diff
@@ -1030,8 +1030,18 @@ def _fixture_teardown(self):
...
                          inhibit_post_migrate=inhibit_post_migrate)
 
-    def assertQuerysetEqual(self, qs, values, transform=repr, ordered=True, msg=None):
+    def assertQuerysetEqual(self, qs, values, transform=None, ordered=True, msg=None):
+        values = list(values)
+
+        if values:
+            if transform is None:
+                if isinstance(values[0], str):
+                    transform = repr
```
Comment (charettes) [inline django/test/testcases.py]: I think we should `warnings.warn('...',  category=RemovedInDjango40Warning)` here.

File: django/test/testcases.py [hunk 4]
```diff
@@ -1030,8 +1030,18 @@ def _fixture_teardown(self):
                          allow_cascade=self.available_apps is not None,
                          inhibit_post_migrate=inhibit_post_migrate)
 
-    def assertQuerysetEqual(self, qs, values, transform=repr, ordered=True, msg=None):
+    def assertQuerysetEqual(self, qs, values, transform=None, ordered=True, msg=None):
+        values = list(values)
+
+        if values:
+            if transform is None:
+                if isinstance(values[0], str):
```
Comment (charettes) [inline django/test/testcases.py]: If we do it the other way around by using an identity transformer `if isinstance(values[0], qs.model)` we would work around the `QuerySet.values_list('charfield', flat=True)` use case.

File: docs/topics/testing/tools.txt [hunk 5]
```diff
@@ -1651,14 +1651,17 @@ your test suite.
...
-    The comparison of the contents of ``qs`` and ``values`` is performed using
-    the function ``transform``; by default, this means that the ``repr()`` of
-    each value is compared. Any other callable can be used if ``repr()`` doesn't
-    provide a unique or helpful comparison.
+    If ``transform`` is supplied as a function, ``values`` is compared to a
+    list produced by applying ``transform`` to each member of ``qs``.
+
+    If ``transform`` is not supplied and ``values`` is a list of ``str``
+    elements, ``values`` is compared to a list produced by applying ``repr()``
+    to each member of ``qs``.
```
Comment (charettes) [inline docs/topics/testing/tools.txt]: I suggest we do it the other way around. If a `values` is a not list of `Model` instances then `repr` is applied to each member of `qs`. I also think we should deprecate this behaviour.

File: django/test/testcases.py [hunk 6]
```diff
@@ -1030,8 +1030,18 @@ def _fixture_teardown(self):
                          allow_cascade=self.available_apps is not None,
                          inhibit_post_migrate=inhibit_post_migrate)
 
-    def assertQuerysetEqual(self, qs, values, transform=repr, ordered=True, msg=None):
+    def assertQuerysetEqual(self, qs, values, transform=None, ordered=True, msg=None):
+        values = list(values)
+
+        if values:
+            if transform is None:
```
Comment (charettes) [inline django/test/testcases.py]: These two could be `and`ed.

File: docs/topics/testing/tools.txt (hunk 5 above)
Comment (inglesp) [reply docs/topics/testing/tools.txt]: We only want to use `repr` if `values` is a list of `str`s.

File: django/test/testcases.py (hunk 4 above)
Comment (inglesp) [reply django/test/testcases.py]: I hadn't thought of flattened values lists.  Have updated the code for this (I think it's a little bit more complicated than just doing the `if` test the other way round) and I've added a test.

**Comment by inglesp** [comment]
Hi @charettes, thanks for your detailed feedback.  Have updated the patch accordingly.

**Comment by jacobtylerwalls** [comment]
LGTM. @inglesp, would you be up for resolving the conflicts and unchecking patch needs improvement on the ticket? Thanks for your work on this. --Jacob

**Comment by inglesp** [comment]
Thanks for the nudge -- I'll try to get to this later this week.

**Comment by inglesp** [comment]
Have rebased (and changed the `RemovedInDjango40Warning` to `RemovedInDjango41Warning`) and now hundreds of test cases are failing with `RemovedInDjango41Warning`.

Fixing these will be tedious, but I'll have a go over the next few days.

**Comment by pope1ni** [comment]
Superseded by #13557.
